### DISCUSSION TITLE
Feature flags

### DISCUSSION THREAD

**Discussion by unixslayer** [ecotoneframework/ecotone-dev#270] [Ideas]
The idea is to be able to define a message endpoint that is accessible but not executable.

Imagine working on a feature that should not be able to change a state in our model, but some parts of the system may already send messages to it. Those messages may be persisted in a queue for later processing.

Ecotone provides API to define Environment Specific Configuration which lets us define an endpoint per environment.

```php
#[Aggregate]
final class MyAggregate
{
    #[CommandHandler]
    #[Environment(['prod'])]
    public function doSomething(DoSomething $command): void
    {
        // business
    }
}
```
However, when an endpoint is defined for a specific environment it is only accessible in that environment and will cause Ecotone to raise an exception of sending a message to an undefined endpoint.

The expected configuration may look as follows

```php
#[Aggregate]
final class MyAggregate
{
    #[CommandHandler]
    #[FeatureFlag('feature-id')]
    public function doSomething(DoSomething $command): void
    {
        // business
    }
}
```

The expected behavior is that other parts of the system may be already sending messages to that endpoint but it won't be executed. For synchronous execution will accept the message but execute it by `NullExecutor`. For asynchronous execution message will be sent to a queue, however running consumer won't be able to run or won't take messages from its queue.

In the case of an asynchronous endpoint by the time the feature gets enabled, it may be already provided with a set of messages that were queued over time.

Setting feature flags for Ecotone may be provided by `ConfigurationVariableService` or `ExtensionObject` when provided by dedicated module.

**Comment by dgafka** [comment]
We have message filters as lower level API, we could levarage that for this. Matter of exposing it as form of interceptor :)

**Comment by unixslayer** [comment]
You're right. I can achieve that with already available components.
